File: remote/src/main/java/com/kelsos/mbrc/services/api/PlaylistService.java
package com.kelsos.mbrc.services.api;

import com.kelsos.mbrc.dto.BaseResponse;
import com.kelsos.mbrc.dto.PaginatedResponse;
import com.kelsos.mbrc.dto.playlist.PlaylistDto;
import com.kelsos.mbrc.dto.playlist.PlaylistTrack;
import com.kelsos.mbrc.dto.playlist.PlaylistTrackInfo;
import com.kelsos.mbrc.dto.requests.MoveRequest;
import com.kelsos.mbrc.dto.requests.PlayPathRequest;
import com.kelsos.mbrc.dto.requests.PlaylistRequest;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;
import rx.Observable;
import rx.Single;

public interface PlaylistService {

  @PUT("/playlists/play")
  Observable<BaseResponse> playPlaylist(@Body PlayPathRequest body);

  @PUT("/playlists/{id}/tracks/move")
  Single<BaseResponse> playlistMoveTrack(@Path("id") int id, @Body MoveRequest body);

  @GET("/playlists/{id}/tracks")
  Single<PaginatedResponse<PlaylistTrack>> getPlaylistTracks(@Path("id") Long id, @Query("offset") int offset,
      @Query("limit") int limit, @Query("after") long after);

  @DELETE("/playlists/{id}/tracks")
  Single<BaseResponse> deletePlaylistTrack(@Path("id") int id, @Query("index") int index);

  @DELETE("/playlists/{id}")
  Single<BaseResponse> deletePlaylist(@Path("id") int id);

  @PUT("/playlists")
  Single<BaseResponse> createPlaylist(@Body PlaylistRequest body);

  @GET("/playlists")
  Observable<PaginatedResponse<PlaylistDto>> getPlaylists(@Query("offset") int offset, @Query("limit") int limit,
      @Query("after") long after);

  @PUT("/playlists/{id}/tracks")
  Single<BaseResponse> addTracksToPlaylist(@Path("id") int id, @Body PlaylistRequest body);

  @GET("/playlists/trackinfo")
  Observable<PaginatedResponse<PlaylistTrackInfo>> getPlaylistTrackInfo(@Query("offset") int offset,
      @Query("limit") int limit,
      @Query("after") long after);
}

File: mbrc-lib/src/main/kotlin/com/kelsos/mbrc/services/api/PlaylistService.kt
package com.kelsos.mbrc.services.api

import com.kelsos.mbrc.dto.BaseResponse
import com.kelsos.mbrc.dto.PaginatedResponse
import com.kelsos.mbrc.dto.playlist.PlaylistDto
import com.kelsos.mbrc.dto.playlist.PlaylistTrack
import com.kelsos.mbrc.dto.playlist.PlaylistTrackInfo
import com.kelsos.mbrc.dto.requests.MoveRequest
import com.kelsos.mbrc.dto.requests.PlayPathRequest
import com.kelsos.mbrc.dto.requests.PlaylistRequest
import retrofit2.http.*
import rx.Observable

interface PlaylistService {

    @PUT("/playlists/play")
    fun playPlaylist(@Body body: PlayPathRequest): Observable<BaseResponse>

    @PUT("/playlists/{id}/tracks/move")
    fun playlistMoveTrack(@Path("id") id: Int, @Body body: MoveRequest): Observable<BaseResponse>

    @GET("/playlists/{id}/tracks")
    fun getPlaylistTracks(@Path("id") id: Long, @Query("offset") offset: Int,
                          @Query("limit") limit: Int, @Query("after") after: Long): Observable<PaginatedResponse<PlaylistTrack>>

    @DELETE("/playlists/{id}/tracks")
    fun deletePlaylistTrack(@Path("id") id: Int, @Query("index") index: Int): Observable<BaseResponse>

    @DELETE("/playlists/{id}")
    fun deletePlaylist(@Path("id") id: Int): Observable<BaseResponse>

    @PUT("/playlists")
    fun createPlaylist(@Body body: PlaylistRequest): Observable<BaseResponse>

    @GET("/playlists")
    fun getPlaylists(@Query("offset") offset: Int, @Query("limit") limit: Int,
                     @Query("after") after: Long): Observable<PaginatedResponse<PlaylistDto>>

    @PUT("/playlists/{id}/tracks")
    fun addTracksToPlaylist(@Path("id") id: Int, @Body body: PlaylistRequest): Observable<BaseResponse>

    @GET("/playlists/trackinfo")
    fun getPlaylistTrackInfo(@Query("offset") offset: Int,
                             @Query("limit") limit: Int,
                             @Query("after") after: Long): Observable<PaginatedResponse<PlaylistTrackInfo>>
}
